cpp/src/io/csv/csv_gpu.cu

```diff
@@ -959,116 +959,107 @@ __global__ void __launch_bounds__(rowofs_block_dim) gather_row_offsets_gpu(uint6
-                                                             size_t num_rows,
-                                                             size_t num_actual_columns,
-                                                             size_t num_active_columns,
-                                                             const cudf::io::ParseOptions &options,
-                                                             column_parse::flags *flags,
-                                                             cudaStream_t stream)
+thrust::host_vector<column_parse::stats> detect_column_types(
+  cudf::io::ParseOptions const &options,
+  device_span<char const> const &data,
+  device_span<column_parse::flags const> const column_flags,
+  device_span<uint64_t const> const &row_starts,
+  size_t const num_active_columns,
+  cudaStream_t stream)
 {
   // Calculate actual block count to use based on records count
   const int block_size = csvparse_block_dim;
-  const int grid_size  = (num_rows + block_size - 1) / block_size;
+  const int grid_size  = (row_starts.size() + block_size - 1) / block_size;
 
   auto d_stats = rmm::device_vector<column_parse::stats>(num_active_columns);
 
   data_type_detection<<<grid_size, block_size, 0, stream>>>(
-    data, options, num_rows, num_actual_columns, flags, row_starts, d_stats.data().get());
+    options, data, column_flags, row_starts, d_stats);
 
   return thrust::host_vector<column_parse::stats>(d_stats);
 }
 
-cudaError_t __host__ DecodeRowColumnData(const char *data,
-                                         const uint64_t *row_starts,
-                                         size_t num_rows,
-                                         size_t num_columns,
-                                         const ParseOptions &options,
-                                         const column_parse::flags *flags,
-                                         cudf::data_type *dtypes,
-                                         void **columns,
-                                         cudf::bitmask_type **valids,
-                                         cudaStream_t stream)
+void __host__ decode_row_column_data(cudf::io::ParseOptions const &options,
+                                     device_span<char const> const &data,
+                                     device_span<column_parse::flags const> const &flags,
+                                     device_span<uint64_t const> const &row_offsets,
+                                     cudf::data_type *dtypes,
+                                     void **columns,
+                                     cudf::bitmask_type **valids,
+                                     cudaStream_t stream)
 {
   // Calculate actual block count to use based on records count
   const int block_size = csvparse_block_dim;
-  const int grid_size  = (num_rows + block_size - 1) / block_size;
+  const int grid_size  = (row_offsets.size() + block_size - 1) / block_size;
```

Comment:
```suggestion
  auto const num_rows = row_offsets.size() - 1;
  const int grid_size  = (num_rows + block_size - 1) / block_size;
```

Reply:
Alternatively, you can change the `row_offsets` to exclude the end of the last row and have the same logic as in the JSON reader.